### PR TITLE
[ALLUXIO-3218] Make gaining primacy interruptible

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-3218

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java [hunk 1]
```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
...
+      }
+    }
+  }
+
+  /**
+   * @return true if the master successfully upgrades to primary
+   */
+  private boolean gainPrimacy() throws Exception {
+    AtomicBoolean lostPrimacy = new AtomicBoolean(false);
+    try (Scoped scoped = mLeaderSelector.onStateChange(state -> lostPrimacy.set(true))) {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: does this have to check what the `state` is?

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java [hunk 2]
```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
...
-        } catch (Throwable t) {
-          Throwable root = ExceptionUtils.getRootCause(t);
-          if ((root != null && (root instanceof InterruptedException)) || Thread.interrupted()) {
-            return;
-          }
-          ProcessUtils.fatalError(LOG, t, "Exception thrown in main serving thread");
+      mJournalSystem.gainPrimacy();
+      if (lostPrimacy.get()) {
+        losePrimacy();
+        return false;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: at this point the journal has done `gainPrimacy` but we return `false`? Is that fine?

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java [hunk 3]
```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
...
-        try {
-          startServing(" (gained leadership)", " (lost leadership)");
-        } catch (Throwable t) {
-          Throwable root = ExceptionUtils.getRootCause(t);
-          if ((root != null && (root instanceof InterruptedException)) || Thread.interrupted()) {
-            return;
-          }
-          ProcessUtils.fatalError(LOG, t, "Exception thrown in main serving thread");
+      mJournalSystem.gainPrimacy();
+      if (lostPrimacy.get()) {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: Why do we need to check `lostPrimacy`, and not `mLeaderSelector.getState()`?

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java [hunk 4]
```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
...
       mLeaderSelector.waitForState(State.PRIMARY);
+      if (gainPrimacy()) {
+        mLeaderSelector.waitForState(State.SECONDARY);
+        losePrimacy();
+      }
+    }
+  }
+
+  /**
+   * @return true if the master successfully upgrades to primary
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: I don't fully understand what this method does. Has the `mLeaderSelector` already notified this should be primary now? This method just does the work to transition?

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java [hunk 5]
```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
       throw new RuntimeException(e);
     }
 
+    startMasters(false);
     while (!Thread.interrupted()) {
-      startMasters(false);
-      LOG.info("Secondary started");
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: added back, thanks

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java (hunk 4 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: Right, the main loop is
```java
while (!Thread.interrupted()) {
      mLeaderSelector.waitForState(State.PRIMARY);
      if (gainPrimacy()) {
        mLeaderSelector.waitForState(State.SECONDARY);
        losePrimacy();
      }
    }
```
so we only call this method after the leader selector decides that it is primary, and this method does the work to transition. I've updated the javadoc and added a comment to give more explanation

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java (hunk 1 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: master should only complete the upgrade to primary if it stays primary throughout the time it takes to call `journalSystem.gainPrimacy()`. I've renamed this variable to `unstable` to make it clearer

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java (hunk 3 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: 1. master A becomes primary and begins journal upgrade
2. master A becomes secondary (continues journal upgrade)
3. master B becomes primary and serves some requests
4. master A becomes primary again
5. master A finishes journal upgrade, checks `mLeaderSelector.getState()` and starts serving, but loses the requests that master B served while A was upgrading.

File: core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java (hunk 2 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java]: `losePrimacy` will call `mJournalSystem.losePrimacy()`, so we will be back in secondary mode when we return `false`

**Comment by aaudiber** [comment]
@gpang comments addressed, PTAL

**Comment by aaudiber** [comment]
Jenkins, test this please

**Pull request merged**